packages/eslint-plugin/src/rules/require-types-exports.ts

```diff
@@ -0,0 +1,386 @@
+        node.typeParameters?.params.forEach(param => collect(param.constraint));
+        node.params.forEach(collect);
+        collect(node.returnType?.typeAnnotation);
+        collectFunctionReturnStatements(node).forEach(collect);
+        break;
+
+      case AST_NODE_TYPES.AssignmentPattern:
+        collect(node.left);
+        break;
+
+      case AST_NODE_TYPES.RestElement:
+        collect(node.argument);
+        collect(node.typeAnnotation?.typeAnnotation);
+        break;
+
+      case AST_NODE_TYPES.ObjectPattern:
+        node.properties.forEach(collect);
+        collect(node.typeAnnotation?.typeAnnotation);
+        break;
+
+      case AST_NODE_TYPES.ArrayPattern:
+        node.elements.forEach(collect);
+        collect(node.typeAnnotation?.typeAnnotation);
+
+        break;
+
+      case AST_NODE_TYPES.ReturnStatement:
+        collect(node.argument);
+        break;
+
+      case AST_NODE_TYPES.TSTypeReference: {
+        const scope = sourceCode.getScope(node);
+        const variable = findVariable(scope, getTypeName(node.typeName));
+
+        const isBuiltinType = variable instanceof ImplicitLibVariable;
+
+        const isGenericTypeArg =
+          variable?.scope.type === ScopeType.function &&
+          variable.identifiers.every(
+            id => id.parent.type === AST_NODE_TYPES.TSTypeParameter,
+          );
+
+        if (!isBuiltinType && !isGenericTypeArg) {
+          typeReferences.add(node);
+        }
+
+        node.typeArguments?.params.forEach(collect);
+        break;
+      }
+
+      case AST_NODE_TYPES.TSArrayType:
+        collect(node.elementType);
+        break;
+
+      case AST_NODE_TYPES.TSTupleType:
+        node.elementTypes.forEach(collect);
+        break;
+
+      case AST_NODE_TYPES.TSUnionType:
+      case AST_NODE_TYPES.TSIntersectionType:
+        node.types.forEach(collect);
+        break;
+
+      case AST_NODE_TYPES.TSTypeLiteral:
+        node.members.forEach(collect);
+        break;
+
+      case AST_NODE_TYPES.TSPropertySignature:
+        collect(node.typeAnnotation?.typeAnnotation);
+        break;
+
+      case AST_NODE_TYPES.TSQualifiedName:
+        collect(node.parent);
+        break;
+
+      case AST_NODE_TYPES.TSAsExpression:
+        collect(node.expression);
+        collect(node.typeAnnotation);
+        break;
+
+      case AST_NODE_TYPES.TSIndexedAccessType:
+        collect(node.objectType);
+        collect(node.indexType);
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  return typeReferences;
+}
+
+function collectFunctionReturnStatements(
+  functionNode: TSESTree.Node,
+): Set<TSESTree.Node> {
+  const isArrowFunctionReturn =
+    functionNode.type === AST_NODE_TYPES.ArrowFunctionExpression &&
+    functionNode.body.type === AST_NODE_TYPES.Identifier;
+
+  if (isArrowFunctionReturn) {
+    return new Set([functionNode.body]);
+  }
+
+  const returnStatements = new Set<TSESTree.Node>();
+
+  simpleTraverse(functionNode, {
```

Comment:
[Refactor] I don't think `simpleTraverse` is what we should use here. It's tailored to the way it's currently used in `packages/typescript-estree`:

* There's no way to break, but it looks like this usage would want to halt traversal after finding a `ReturnStatement`
* It allows configurable selectors with `'enter'`, but this code just uses `ReturnStatement`

I think it'd be better to avoid `simpleTraverse` here, and instead do some or all of:
* Look into the existing `forEachReturnStatement`, maybe it's exactly what you need?
* Failing that, write a dedicated simpler traverser for just this rule that halts on `ReturnStatement`s, omitting anything around parent pointers or `'enter'`
* Reuse some of the design and logic from one of the rules that does similar things: `explicit-module-boundary-types`, `explicit-function-return-type`
  * This one I'm low confidence in, but figured I'd mention...

Reply:
Created a custom traverser, inspired by the original `forEachReturnStatement`